File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.js
/**
 * @param {number[]} arr
 * @return {number[]}
 */
var arrayRankTransform = function(arr) {
    let nums = [...arr].sort((a, b) => a - b);

    let rank = 1;
    let hashmap = {};

    for (let i = 0; i < nums.length; i++) {
        if (hashmap[nums[i]]) continue;

        hashmap[nums[i]] = rank;
        rank++;
    }

   return arr.map(num => hashmap[num]);
};

File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        TreeSet<Integer> ts = new TreeSet<>();
        for (int n : arr) {
            ts.add(n);
        }
        HashMap<Integer, Integer> num = new HashMap<>();
        int count = 1;
        for (int n : ts) {
            num.put(n, count++);
        }
        int[] ans = new int[arr.length];
        for (int i = 0; i < arr.length; i++) {
            ans[i] = num.get(arr[i]);
        }
        return ans;
    }
}